src/Workspaces/Core/Portable/SemanticModelReuse/AbstractSemanticModelReuseLanguageService.cs

```diff
@@ -30,14 +32,38 @@ internal abstract class AbstractSemanticModelReuseLanguageService<
         protected abstract SyntaxList<TAccessorDeclarationSyntax> GetAccessors(TBasePropertyDeclarationSyntax baseProperty);
         protected abstract TBasePropertyDeclarationSyntax GetBasePropertyDeclaration(TAccessorDeclarationSyntax accessor);
 
-        public Task<SemanticModel?> TryGetSpeculativeSemanticModelAsync(SemanticModel previousSemanticModel, SyntaxNode currentBodyNode, CancellationToken cancellationToken)
+        public async Task<SemanticModel?> TryGetSpeculativeSemanticModelAsync(SemanticModel previousSemanticModel, SyntaxNode currentBodyNode, CancellationToken cancellationToken)
         {
             var previousSyntaxTree = previousSemanticModel.SyntaxTree;
             var currentSyntaxTree = currentBodyNode.SyntaxTree;
 
-            // This operation is only valid if top-level equivalent trees were passed in.
-            Contract.ThrowIfFalse(previousSyntaxTree.IsEquivalentTo(currentSyntaxTree, topLevel: true));
-            return TryGetSpeculativeSemanticModelWorkerAsync(previousSemanticModel, currentBodyNode, cancellationToken);
+            // This operation is only valid if top-level equivalent trees were passed in.  If they're not equivalent
+            // then something very bad happened as we did that document.Project.GetDependentSemanticVersionAsync was
+            // still the same.  So somehow w don't have top-level equivalence, but we do have the same semantic version.
+            //
+            // log a NFW to help diagnose what the source looks like as it may help us determine what sort of edit is
+            // causing this.
+            if (!previousSyntaxTree.IsEquivalentTo(currentSyntaxTree, topLevel: true))
+            {
+                try
+                {
+                    throw new InvalidOperationException(
```

Comment:
⚠️ I'm a bit concerned about the servicing impact of this exception. I would prefer to have a static latch to ensure we only send this report at most once per VS session, which avoids cascading impact of large numbers of calls to `GetText(...).ToString()`.